File: apps/web/lib/privacy/academy-analytics.ts
/**
 * Academy creator-analytics telemetry for Pulse web.
 *
 * Dual-path:
 * 1) Consent-gated Firebase Analytics custom events (→ BigQuery when linked)
 * 2) Always-on trusted callable `recordAcademyAnalytics` for Studio rollups
 *
 * Payloads never include uid, email, or free text.
 */
import {
  ACADEMY_ANALYTICS_SCHEMA_VERSION,
  ANALYTICS_HEARTBEAT_SECONDS,
  type AcademyAnalyticsEventInput,
  type AcademyAnalyticsEventName,
  type AcademyAnalyticsPlatform,
  type AcademyAnalyticsSource,
} from "@pulse/shared";
import { callCloudFunction } from "@/lib/firebase/call-function";
import { getAnalyticsConsent } from "@/lib/privacy/telemetry";

const SESSION_KEY = "pulse_academy_session_v1";
const QUEUE_FLUSH_MS = 2500;
const MAX_QUEUE = 40;

type Queued = AcademyAnalyticsEventInput & { clientEventId: string };

let queue: Queued[] = [];
let flushTimer: ReturnType<typeof setTimeout> | null = null;
let flushing = false;

function randomId(): string {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID().replace(/-/g, "").slice(0, 24);
  }
  return `${Date.now().toString(36)}${Math.random().toString(36).slice(2, 10)}`;
}

export function getAcademySessionId(): string {
  if (typeof window === "undefined") return "ssr";
  try {
    const existing = window.sessionStorage.getItem(SESSION_KEY);
    if (existing) return existing;
    const next = randomId();
    window.sessionStorage.setItem(SESSION_KEY, next);
    return next;
  } catch {
    return randomId();
  }
}

function platform(): AcademyAnalyticsPlatform {
  return "web";
}

function gaEventName(name: AcademyAnalyticsEventName): string {
  return `academy_${name}`;
}

async function logToGa(event: AcademyAnalyticsEventInput): Promise<void> {
  if (typeof window === "undefined") return;
  if (!getAnalyticsConsent()) return;
  const measurementId = process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID;
  if (!measurementId) return;
  try {
    const { getAnalytics, isSupported, logEvent } = await import(
      "firebase/analytics"
    );
    if (!(await isSupported())) return;
    const { getFirebaseApp } = await import("@/lib/firebase/client");
    const analytics = getAnalytics(getFirebaseApp());
    logEvent(analytics, gaEventName(event.name), {
      course_id: event.courseId,
      lesson_id: event.lessonId ?? undefined,
      source: event.source ?? "unknown",
      platform: event.platform,
      watch_delta: event.watchDeltaSeconds ?? 0,
      retention_bucket: event.retentionBucket ?? undefined,
      traffic_source: event.trafficSource ?? undefined,
      schema_version: ACADEMY_ANALYTICS_SCHEMA_VERSION,
    });
  } catch {
    // Best-effort.
  }
}

async function flushQueue(): Promise<void> {
  if (flushing || queue.length === 0) return;
  flushing = true;
  const batch = queue.splice(0, 20);
  try {
    await callCloudFunction("recordAcademyAnalytics", { events: batch });
  } catch {
    // Re-queue a limited number so transient failures don't drop everything.
    queue = [...batch.slice(0, 10), ...queue].slice(0, MAX_QUEUE);
  } finally {
    flushing = false;
    if (queue.length > 0) scheduleFlush();
  }
}

function scheduleFlush(): void {
  if (flushTimer) return;
  flushTimer = setTimeout(() => {
    flushTimer = null;
    void flushQueue();
  }, QUEUE_FLUSH_MS);
}

export function trackAcademyEvent(
  partial: Omit<AcademyAnalyticsEventInput, "sessionId" | "platform"> & {
    sessionId?: string;
    platform?: AcademyAnalyticsPlatform;
  },
): void {
  if (typeof window === "undefined") return;
  const event: Queued = {
    ...partial,
    sessionId: partial.sessionId ?? getAcademySessionId(),
    platform: partial.platform ?? platform(),
    source: (partial.source ?? "unknown") as AcademyAnalyticsSource,
    clientEventId: randomId(),
  };
  void logToGa(event);
  queue.push(event);
  if (queue.length >= 15) {
    void flushQueue();
  } else {
    scheduleFlush();
  }
}

export function trackCourseImpression(input: {
  courseId: string;
  source?: AcademyAnalyticsSource;
  trafficSource?: string | null;
}): void {
  trackAcademyEvent({
    name: "course_impression",
    courseId: input.courseId,
    source: input.source,
    trafficSource: input.trafficSource,
  });
}

export function trackCourseOpen(input: {
  courseId: string;
  source?: AcademyAnalyticsSource;
  trafficSource?: string | null;
  locale?: string;
}): void {
  trackAcademyEvent({
    name: "course_open",
    courseId: input.courseId,
    source: input.source,
    trafficSource: input.trafficSource,
    locale: input.locale,
  });
}

export function trackLessonStart(input: {
  courseId: string;
  lessonId: string;
  durationSeconds?: number;
  source?: AcademyAnalyticsSource;
}): void {
  trackAcademyEvent({
    name: "lesson_start",
    courseId: input.courseId,
    lessonId: input.lessonId,
    durationSeconds: input.durationSeconds,
    source: input.source,
  });
}

export function trackLessonHeartbeat(input: {
  courseId: string;
  lessonId: string;
  positionSeconds: number;
  durationSeconds: number;
  watchDeltaSeconds: number;
}): void {
  const duration = Math.max(1, input.durationSeconds);
  const bucket = Math.min(
    100,
    Math.max(0, Math.floor((input.positionSeconds / duration) * 100)),
  );
  trackAcademyEvent({
    name: "lesson_heartbeat",
    courseId: input.courseId,
    lessonId: input.lessonId,
    positionSeconds: input.positionSeconds,
    durationSeconds: input.durationSeconds,
    watchDeltaSeconds: input.watchDeltaSeconds,
    retentionBucket: bucket,
  });
}

export function trackLessonComplete(input: {
  courseId: string;
  lessonId: string;
  positionSeconds?: number;
  durationSeconds?: number;
}): void {
  trackAcademyEvent({
    name: "lesson_complete",
    courseId: input.courseId,
    lessonId: input.lessonId,
    positionSeconds: input.positionSeconds,
    durationSeconds: input.durationSeconds,
    retentionBucket: 100,
  });
}

export function trackQuizSubmit(input: {
  courseId: string;
  lessonId: string;
  passed: boolean;
  score: number;
}): void {
  trackAcademyEvent({
    name: "quiz_submit",
    courseId: input.courseId,
    lessonId: input.lessonId,
    quizPassed: input.passed,
    quizScore: input.score,
  });
}

export { ANALYTICS_HEARTBEAT_SECONDS };

File: apps/mobile/lib/privacy/academy_analytics.dart
import 'dart:async';
import 'dart:math';

import 'package:firebase_analytics/firebase_analytics.dart';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../firebase/https_callable.dart';
import 'telemetry.dart';

/// Schema version mirrored from `@pulse/shared` ACADEMY_ANALYTICS_SCHEMA_VERSION.
const academyAnalyticsSchemaVersion = 1;

/// Heartbeat interval expected by the Functions aggregator.
const academyAnalyticsHeartbeat = Duration(seconds: 15);

const _sessionKey = 'pulse_academy_session_v1';

final _queue = <Map<String, dynamic>>[];
Timer? _flushTimer;
bool _flushing = false;
String? _sessionId;

String _randomId() {
  final rand = Random.secure();
  final bytes = List<int>.generate(12, (_) => rand.nextInt(256));
  return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
}

Future<String> academySessionId() async {
  if (_sessionId != null) return _sessionId!;
  final prefs = await SharedPreferences.getInstance();
  final existing = prefs.getString(_sessionKey);
  if (existing != null && existing.isNotEmpty) {
    _sessionId = existing;
    return existing;
  }
  final next = _randomId();
  await prefs.setString(_sessionKey, next);
  _sessionId = next;
  return next;
}

String _platform() {
  if (kIsWeb) return 'web';
  switch (defaultTargetPlatform) {
    case TargetPlatform.iOS:
      return 'ios';
    case TargetPlatform.android:
      return 'android';
    default:
      return 'web';
  }
}

Future<void> _logGa(String name, Map<String, Object?> params) async {
  if (!await TelemetryPrefs.isAnalyticsEnabled()) return;
  try {
    await FirebaseAnalytics.instance.logEvent(
      name: 'academy_$name',
      parameters: {
        for (final entry in params.entries)
          if (entry.value != null) entry.key: entry.value!,
        'schema_version': academyAnalyticsSchemaVersion,
      },
    );
  } catch (_) {
    // Best-effort.
  }
}

void _scheduleFlush() {
  _flushTimer ??= Timer(const Duration(milliseconds: 2500), () {
    _flushTimer = null;
    unawaited(_flushQueue());
  });
}

Future<void> _flushQueue() async {
  if (_flushing || _queue.isEmpty) return;
  _flushing = true;
  final batch = _queue.take(20).toList(growable: false);
  _queue.removeRange(0, batch.length);
  try {
    await HttpsCallableClient().call('recordAcademyAnalytics', <String, dynamic>{
      'events': batch,
    });
  } catch (e) {
    debugPrint('[AcademyAnalytics] flush failed: $e');
    _queue.insertAll(0, batch.take(10));
    if (_queue.length > 40) {
      _queue.removeRange(40, _queue.length);
    }
  } finally {
    _flushing = false;
    if (_queue.isNotEmpty) _scheduleFlush();
  }
}

Future<void> trackAcademyEvent({
  required String name,
  required String courseId,
  String? lessonId,
  String source = 'unknown',
  int? positionSeconds,
  int? durationSeconds,
  int? watchDeltaSeconds,
  int? retentionBucket,
  bool? quizPassed,
  int? quizScore,
  String? locale,
  String? trafficSource,
}) async {
  final sessionId = await academySessionId();
  final event = <String, dynamic>{
    'name': name,
    'courseId': courseId,
    'sessionId': sessionId,
    'platform': _platform(),
    'source': source,
    'clientEventId': _randomId(),
    'lessonId': ?lessonId,
    'positionSeconds': ?positionSeconds,
    'durationSeconds': ?durationSeconds,
    'watchDeltaSeconds': ?watchDeltaSeconds,
    'retentionBucket': ?retentionBucket,
    'quizPassed': ?quizPassed,
    'quizScore': ?quizScore,
    'locale': ?locale,
    'trafficSource': ?trafficSource,
  };
  await _logGa(name, {
    'course_id': courseId,
    'lesson_id': lessonId,
    'source': source,
    'platform': _platform(),
    'watch_delta': watchDeltaSeconds ?? 0,
    'retention_bucket': retentionBucket,
    'traffic_source': trafficSource,
  });
  _queue.add(event);
  if (_queue.length >= 15) {
    unawaited(_flushQueue());
  } else {
    _scheduleFlush();
  }
}

Future<void> trackCourseImpression({
  required String courseId,
  String source = 'catalog',
}) {
  return trackAcademyEvent(
    name: 'course_impression',
    courseId: courseId,
    source: source,
  );
}

Future<void> trackCourseOpen({
  required String courseId,
  String source = 'direct',
  String? locale,
}) {
  return trackAcademyEvent(
    name: 'course_open',
    courseId: courseId,
    source: source,
    locale: locale,
  );
}

Future<void> trackLessonStart({
  required String courseId,
  required String lessonId,
  int? durationSeconds,
}) {
  return trackAcademyEvent(
    name: 'lesson_start',
    courseId: courseId,
    lessonId: lessonId,
    durationSeconds: durationSeconds,
  );
}

Future<void> trackLessonHeartbeat({
  required String courseId,
  required String lessonId,
  required int positionSeconds,
  required int durationSeconds,
  required int watchDeltaSeconds,
}) {
  final duration = durationSeconds <= 0 ? 1 : durationSeconds;
  final bucket = ((positionSeconds / duration) * 100).floor().clamp(0, 100);
  return trackAcademyEvent(
    name: 'lesson_heartbeat',
    courseId: courseId,
    lessonId: lessonId,
    positionSeconds: positionSeconds,
    durationSeconds: durationSeconds,
    watchDeltaSeconds: watchDeltaSeconds,
    retentionBucket: bucket,
  );
}

Future<void> trackLessonComplete({
  required String courseId,
  required String lessonId,
  int? positionSeconds,
  int? durationSeconds,
}) {
  return trackAcademyEvent(
    name: 'lesson_complete',
    courseId: courseId,
    lessonId: lessonId,
    positionSeconds: positionSeconds,
    durationSeconds: durationSeconds,
    retentionBucket: 100,
  );
}

Future<void> trackQuizSubmit({
  required String courseId,
  required String lessonId,
  required bool passed,
  required int score,
}) {
  return trackAcademyEvent(
    name: 'quiz_submit',
    courseId: courseId,
    lessonId: lessonId,
    quizPassed: passed,
    quizScore: score,
  );
}
